presto-orc/src/main/java/com/facebook/presto/orc/reader/MapFlatSelectiveStreamReader.java

```diff
@@ -0,0 +1,830 @@
+        implements SelectiveStreamReader
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(MapFlatSelectiveStreamReader.class).instanceSize();
+
+    private final StreamDescriptor streamDescriptor;
+
+    // This is the StreamDescriptor for the value stream with sequence ID 0, it is used to derive StreamDescriptors for the
+    // value streams with other sequence IDs
+    private final StreamDescriptor baseValueStreamDescriptor;
+    private final OrcTypeKind keyOrcType;
+    private final DateTimeZone hiveStorageTimeZone;
+
+    private final boolean nullsAllowed;
+    private final boolean nonNullsAllowed;
+    private final boolean outputRequired;
+    @Nullable
+    private final MapType outputType;
+    @Nullable
+    private final Set<Long> requiredLongKeys;
+    @Nullable
+    private final Set<String> requiredStringKeys;
+
+    private int[] keyIndices;   // indices of the required keys in the encodings array passed to startStripe
+    private int keyCount;       // number of valid entries in keyIndices array
+
+    private final List<InputStreamSource<BooleanInputStream>> inMapStreamSources = new ArrayList<>();
+    private final List<BooleanInputStream> inMapStreams = new ArrayList<>();
+    private final List<SelectiveStreamReader> valueStreamReaders = new ArrayList<>();
+    private final List<StreamDescriptor> valueStreamDescriptors = new ArrayList<>();
+
+    private Block keyBlock;
+    private int readOffset;
+    private int[] nestedReadOffsets;
+
+    private InputStreamSource<BooleanInputStream> presentStreamSource = missingStreamSource(BooleanInputStream.class);
+    @Nullable
+    private BooleanInputStream presentStream;
+
+    private boolean rowGroupOpen;
+    private int[] offsets;
+    private boolean[] nulls;
+    private int[] outputPositions;
+    private int outputPositionCount;
+    private boolean outputPositionsReadOnly;
+    private boolean allNulls;
+    private boolean valuesInUse;
+
+    private int[] nestedLengths;
+    private int[][] nestedPositions;
+    private int[] nestedPositionCounts;
+    private int[][] nestedOutputPositions;
+    private boolean[][] inMap;
+
+    private final AggregatedMemoryContext systemMemoryContext;
+    private final LocalMemoryContext localMemoryContext;
+
+    public MapFlatSelectiveStreamReader(
+            StreamDescriptor streamDescriptor,
+            Map<Subfield, TupleDomainFilter> filters,
+            List<Subfield> requiredSubfields,
+            Optional<Type> outputType,
+            DateTimeZone hiveStorageTimeZone,
+            AggregatedMemoryContext systemMemoryContext)
+    {
+        checkArgument(filters.keySet().stream().map(Subfield::getPath).allMatch(List::isEmpty), "filters on nested columns are not supported yet");
+
+        this.streamDescriptor = requireNonNull(streamDescriptor, "streamDescriptor is null");
+        this.keyOrcType = streamDescriptor.getNestedStreams().get(0).getOrcTypeKind();
```

Comment:
can you rename this to "keyOrcTypeKind". also do we need to check the size of the nested stream before?

Reply:
@bhhari Will make this change:

```
        checkArgument(streamDescriptor.getNestedStreams().size() == 2, "there must be exactly 2 nested stream descriptor");
        this.keyOrcTypeKind = streamDescriptor.getNestedStreams().get(0).getOrcTypeKind();
```